src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -156,11 +169,63 @@ public override void VisitMarkupTagHelperAttribute(MarkupTagHelperAttributeSynta
             {
                 if (node.TagHelperAttributeInfo.Bound)
                 {
-                    _syntaxNodes.Add(node.Name);
+                    _syntaxNodes.Add((node.Name, SyntaxKind.MarkupTagHelperAttribute));
                 }
 
                 base.VisitMarkupTagHelperAttribute(node);
             }
+
+            public override void VisitMarkupTagHelperDirectiveAttribute(MarkupTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    _syntaxNodes.Add((node.Transition, SyntaxKind.Transition));
+                    _syntaxNodes.Add((node.Name, SyntaxKind.MarkupTagHelperDirectiveAttribute));
```

Comment:
My reasoning is that once you do `node.Name`, node.Kind is something like `MarkupTextLiteral`, which is true of TagHelperElement, TagHelperAttribute, DirectiveAttribute and DirectiveAttributeParameter (and DirectiveAttributes need to be distinguishable from THElements and THAttributes). You could inspect the node.Parent (or pass the whole `MarkupTagHelperDirectiveAttribute` node), but then we'd be putting logic about the structure of the tree outside of the Visitor, and that feels fragile/the wrong separation to me. I'm also not a fan or returning Tuples, but it was the cleanest idea I had. Maybe instead of returning `(SyntaxNode, SyntaxKind)` it should be `(Range, SyntaxKind)` or something similar? That way it's only given the parts it should use.

Reply:
@ajaybhargavb what do you think?

Reply:
> Maybe instead of returning (SyntaxNode, SyntaxKind) it should be (Range, SyntaxKind) or something similar?

I like this. In fact, if you move the logic in the switch statement above into this visitor, this visitor can just return an array of uints directly. It'll also keep you from modifying two places(visitor and switch statement) if we change the logic/add more ranges in the future. More importantly, I like keeping all the syntax knowledge contained in the visitor.

Reply:
@ajaybhargavb the Visitor returning IList<uint> would be nice, but since the uint results of the current node are dependent on the previous node its easier to calculate everything if they're still stored as Nodes/Ranges and then we make a second pass.